src/mgxs.cpp

```diff
@@ -575,24 +573,20 @@ Mgxs::sample_fission_energy(int gin, int& dg, int& gout)
     // the neutron is delayed
 
     // get the delayed group
-    dg = 0;
-    while (xi_pd >= prob_prompt) {
-      dg++;
-      prob_prompt +=
-           xs_t->delayed_nu_fission(cache[tid].a, dg, gin);
+    for (dg = 0; dg < num_delayed_groups; ++dg) {
+      prob_prompt += xs_t->delayed_nu_fission(cache[tid].a, dg, gin);
+      if (xi_pd < prob_prompt) break;
     }
 
     // adjust dg in case of round-off error
     dg = std::min(dg, num_delayed_groups - 1);
 
     // sample the outgoing energy group
     gout = 0;
-    double prob_gout =
-         xs_t->chi_delayed(cache[tid].a, dg, gin, gout);
+    double prob_gout = xs_t->chi_delayed(cache[tid].a, dg, gin, gout);
     while (prob_gout < xi_gout) {
       gout++;
```

Comment:
Note to self: when fixing that group variables are still 1-based, fix this otherwise will get same bug as above.